src/frequenz/core/asyncio/_task_group.py

```diff
@@ -0,0 +1,392 @@
+        [`as_completed()`][frequenz.core.asyncio.PersistentTaskGroup.as_completed]
+        should be used.
+
+        Args:
+            coro: The coroutine to be managed.
+            name: The name of the task. Names will always have the form
+                `f"{self}:{name}"`. If `None` or empty, the default name will be
+                `hex(id(coro))[2:]`. If you need the final name of the task, it can
+                always be retrieved
+            context: The context to be used for the task.
+            log_exception: Whether to log exceptions raised by the task.
+
+        Returns:
+            The new task.
+        """
+        if not name:
+            name = hex(id(coro))[2:]
+        task = self._task_creator.create_task(
+            coro, name=f"{self}:{name}", context=context
+        )
+        self._running.add(task)
+        task.add_done_callback(self._running.discard)
+        task.add_done_callback(self._waiting_ack.add)
+
+        if log_exception:
+
+            def _log_exception(task: asyncio.Task[TaskReturnT]) -> None:
+                try:
+                    task.result()
+                except asyncio.CancelledError:
+                    pass
+                except BaseException:  # pylint: disable=broad-except
+                    _logger.exception(
+                        "Task %s raised an unhandled exception", task.get_name()
+                    )
+
+            task.add_done_callback(_log_exception)
+        return task
+
+    def cancel(self, msg: str | None = None) -> None:
+        """Cancel all running tasks spawned by this group.
+
+        Args:
+            msg: The message to be passed to the tasks being cancelled.
+        """
+        for task in self._running:
+            task.cancel(msg)
+
+    async def stop(self, msg: str | None = None) -> None:
+        """Stop this task group.
+
+        This method cancels all running tasks spawned by this group and waits for them
+        to finish.
+
+        Args:
+            msg: The message to be passed to the tasks being cancelled.
+
+        Raises:
+            BaseExceptionGroup: If any of the tasks spawned by this group raised an
+                exception.
+        """
+        self.cancel(msg)
+        try:
+            await self
+        except BaseExceptionGroup as exc_group:
+            # We want to ignore CancelledError here as we explicitly cancelled all the
+            # tasks.
+            _, rest = exc_group.split(asyncio.CancelledError)
+            if rest is not None:
+                # We are filtering out from an exception group, we really don't want to
+                # add the exceptions we just filtered by adding a from clause here.
+                raise rest  # pylint: disable=raise-missing-from
+
+    async def as_completed(
+        self, *, timeout: float | datetime.timedelta | None = None
+    ) -> AsyncIterator[asyncio.Task[Any]]:
+        """Iterate over completed tasks.
```

Comment:
`Iterate over each running task once its complete.`

I find that more helpful as a headline here, otherwise one might just read the first line and thinks `running` but not complete tasks are ignored

Reply:
True, maybe `Iterate over running tasks yielding as they complete`?

Reply:
Updated